public/app/features/alerting/unified/components/rules/state-history/LokiStateHistory.tsx

```diff
@@ -0,0 +1,350 @@
+      : 'No state transitions have occurred in the last 60 minutes';
+
+  return (
+    <div className={styles.fullSize}>
+      <form onSubmit={handleSubmit((data) => setInstancesFilter(data.query))}>
+        <Field
+          label={
+            <Label htmlFor="instancesSearchInput">
+              <Stack gap={0.5}>
+                <span>Filter instances</span>
+                <HoverCard
+                  content={
+                    <>
+                      Use label matcher expression (like <code>{'{foo=bar}'}</code>) or click on an instance label to
+                      filter instances
+                    </>
+                  }
+                >
+                  <Icon name="info-circle" size="sm" />
+                </HoverCard>
+              </Stack>
+            </Label>
+          }
+        >
+          <Input
+            {...register('query')}
+            id="instancesSearchInput"
+            prefix={<Icon name="search" />}
+            suffix={
+              <Button fill="text" icon="times" size="sm" onClick={onFilterCleared}>
+                Clear
+              </Button>
+            }
+            placeholder="Filter instances"
+          />
+        </Field>
+        <input type="submit" hidden />
+      </form>
+      {!isEmpty(commonLabels) && (
+        <Stack direction="row" alignItems="center">
+          <strong>Common labels</strong>
+          <TagList tags={commonLabels.map((label) => label.join('='))} />
+        </Stack>
+      )}
+      {isEmpty(frameSubset) ? (
+        <>
+          <div className={styles.emptyState}>
+            {emptyStateMessage}
+            {totalRecordsCount > 0 && (
+              <Button variant="secondary" type="button" onClick={onFilterCleared}>
+                Clear filters
+              </Button>
+            )}
+          </div>
+        </>
+      ) : (
+        <>
+          <div className={styles.graphWrapper}>
+            <LogTimelineViewer frames={frameSubset} timeRange={timeRange} onPointerMove={onTimelinePointerMove} />
+          </div>
+          {hasMoreInstances && (
+            <div className={styles.moreInstancesWarning}>
+              <Stack direction="row" alignItems="center" gap={1}>
+                <Icon name="exclamation-triangle" size="sm" />
+                <small>{`Only showing ${frameSubset.length} out of ${dataFrames.length} instances. Click on the labels to narrow down the results`}</small>
+              </Stack>
+            </div>
+          )}
+          <LogRecordViewerByTimestamp
+            records={historyRecords}
+            commonLabels={commonLabels}
+            logsRef={logsRef}
+            onLabelClick={onLogRecordLabelClick}
+          />
+        </>
+      )}
+    </div>
+  );
+};
+
+function useInstanceHistoryRecords(stateHistory?: DataFrameJSON, filter?: string) {
+  const theme = useTheme2();
+
+  return useMemo(() => {
+    // merge timestamp with "line"
+    // @ts-ignore
```

Comment:
This is pretty much the only ts-ignore we have added, do we want to look in to how to type this?